clients/python-wrapper/lakefs/reference.py

```diff
@@ -1,22 +1,179 @@
 
 from lakefs.client import Client, DefaultClient
+from lakefs.exceptions import api_exception_handler
+from lakefs.object import Object
+from lakefs.object_manager import ObjectManager
+
+
+class Commit(NamedTuple):
+    """
+    NamedTuple representing a lakeFS commit's properties
+    """
+    id: str
+    parents: [str]
+    committer: str
+    message: str
+    creation_date: int
+    meta_range_id: str
+    metadata: Optional[dict[str, str]] = None
+
+
+class Change(NamedTuple):
+    """
+    NamedTuple representing a diff change between to refs in lakeFS
+    """
+    type: Literal["added", "removed", "changed", "conflict", "prefix_changed"]
+    path: str
+    path_type: Literal["common_prefix", "object"]
+    size_bytes: Optional[int]
 
 
 class Reference:
     """
     Class representing a reference in lakeFS.
     """
     _client: Client
-    repo_id: str
-    id: str
+    _repo_id: str
+    _id: str
+    _commit: Optional[Commit] = None
 
     def __init__(self, repo_id: str, ref_id: str, client: Optional[Client] = DefaultClient) -> None:
         self._client = client
-        self.repo_id = repo_id
-        self.id = ref_id
+        self._repo_id = repo_id
+        self._id = ref_id
+
+    @property
+    def repo_id(self) -> str:
+        """
+        Return the repository id for this reference
+        """
+        return self._repo_id
+
+    @property
+    def id(self) -> str:
+        """
+        Returns the reference id
+        """
+        return self._id
+
+    @property
+    def objects(self) -> ObjectManager:
+        """
+        Returns a ObjectManager object for this reference
+        """
+        # TODO: Implement
+
+    @staticmethod
+    def _get_generator(func, *args, max_amount: Optional[int] = None, **kwargs):
+        count = 0
+        has_more = True
+        while has_more:
+            with api_exception_handler():
+                page = func(*args, **kwargs)
+                has_more = page.pagination.has_more
+                for res in page.results:
+                    count += 1
+                    yield res
+                    if max_amount is not None and count >= max_amount:
+                        return
+
+    def log(self, max_amount: Optional[int] = None, **kwargs) -> Generator[lakefs_sdk.Commit]:
+        """
+        Returns a generator of commits starting with this reference id
+        :param max_amount: (Optional) limits the amount of results to return from the server
+        :param kwargs: Additional keyword arguments
+        :raises
+            NotFoundException if reference by this id does not exist
+            NotAuthorizedException if user is not authorized to perform this operation
+            ServerException for any other errors
+        """
+        limit = max_amount is not None
+        return self._get_generator(self._client.sdk_client.refs_api.log_commits,
+                                   self._repo_id, self._id, max_amount=max_amount, limit=limit, **kwargs)
```

Comment:
max_amount is a wrapper param - it does not have to adhere to the SDK params. The purpose of the wrapper was to simplify the API. This is exactly it

Reply:
1. Can be very confusion as limit and amount arguments have a special case for log operations. So in case you like to get the best performance from `log` for max_amount you will like to limit the result.
2. Suggest to remove `max_amount` if we use it just to locally. Let the develop stop the loop as required. The caller can count or use https://docs.python.org/2/library/itertools.html#itertools.islice to limit the number of elements.

Reply:
We're trying to simplify the user's code - so adding a max_amount optional parameter seems logical. The advanced user can always ignore this param and provide API parameter via kwargs if needed

Reply:
I believe this value is a language feature rather than an API feature, unless it is used to optimize server calls or is required in some cases when requesting `log` for blame.

@ozkatz, I noticed that this argument is used in different APIs in the proposal. I think we should remove it because Python provides several ways to limit listings.